File: modules/module_sample/src/main/java/com/example/william/my/module/sample/presenter/ArticlePresenter.java
package com.example.william.my.module.sample.presenter;

import com.example.william.my.bean.data.ArticleDetailBean;
import com.example.william.my.module.sample.contract.ArticleContract;
import com.example.william.my.module.sample.repo.ArticleDataSource;
import com.example.william.my.module.sample.repo.ArticleRepository;

import java.util.List;

public class ArticlePresenter implements ArticleContract.Presenter {

    private int mPage;

    private final ArticleRepository mArticleRepository;

    private final ArticleContract.View mArticleView;

    public ArticlePresenter(ArticleRepository articleRepository, ArticleContract.View view) {
        mArticleRepository = articleRepository;
        mArticleView = view;

        this.mArticleView.setPresenter(this);
    }

    @Override
    public void onRefreshArticleList() {
        mPage = 0;
        queryArticleList(mPage);
    }

    @Override
    public void onLoadMoreArticleList() {
        mPage++;
        queryArticleList(mPage);
    }

    private void queryArticleList(int page) {
        mArticleRepository.getArticleList(page, new ArticleDataSource.LoadArticleCallback() {
            @Override
            public void showLoading() {
                mArticleView.showLoading();
            }

            @Override
            public void onArticleLoaded(List<ArticleDetailBean> articles) {
                mArticleView.showArticles(mPage, articles);
            }

            @Override
            public void onDataNotAvailable() {
                if (mPage == 0) {
                    mArticleView.showEmptyView();
                } else {
                    mArticleView.showArticlesNoMore();
                }
            }

            @Override
            public void onFailure(String msg) {
                if (mPage == 0) {
                    mArticleView.showEmptyView();
                }
                mArticleView.showToast(msg);
            }
        });
    }


    @Override
    public void start() {

    }

    @Override
    public void clear() {

    }
}

File: modules/module_sample/src/main/kotlin/com/example/william/my/module/sample/presenter/ArticlePresenter.kt
package com.example.william.my.module.sample.presenter

import com.example.william.my.bean.data.ArticleDetailBean
import com.example.william.my.bean.repo.DataRepository
import com.example.william.my.bean.repo.DataSource.LoadArticleCallback
import com.example.william.my.module.sample.contract.ArticleContract

class ArticlePresenter(private val mDataRepository: DataRepository, private val mArticleView: ArticleContract.View) : ArticleContract.Presenter {

    override fun queryArticleList(page: Int) {
        mDataRepository.queryArticleList(page, object : LoadArticleCallback {
            override fun showLoading() {

            }

            override fun onFailure(msg: String) {
                mArticleView.onDataFail()
                mArticleView.showToast(msg)
            }

            override fun onArticleLoaded(articles: List<ArticleDetailBean>) {
                mArticleView.showArticles(articles)
            }
        })
    }

    override fun start() {

    }

    override fun clear() {

    }
}